### PR TITLE
nixos/borgbackup: fix extraArgs shell expansion

### LINKED ISSUE #323262: borgbackup: extraCreateArgs (and other args) are broken
### Describe the bug

`extraCreateArgs` escapes the args, breaking them because they are not expanded by bash, being passed from systemd through an env var. Easier to understand with the repro:

### Steps To Reproduce
1. Set up a borgbackup job `test` with `extraCreateArgs = [ "--exclude-caches" ]`
2. `systemctl cat borgbackup-job-test` will have a line `Environment="extraCreateArgs='--exclude-caches'"` 
3. `systemctl start borgbackup-job-test` will then fail with `borg create: error: argument ARCHIVE: "'--exclude-caches'": No archive specified`
4. Manually setting extraCreateArgs correctly in the `preHook` works around the issue.

### Expected behavior
It should work as expected, I assume the same issue applies to other extraArgs parameters.

### Notify maintainers

@dotlambda
<!--

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
λ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.9, NixOS, 24.11 (Vicuna), 24.11.20240624.2893f56`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.23.0`
 - channels(necauqua): `""`
 - channels(root): `""`
 - nixpkgs: `/nix/store/vsbi6gfwf2bdamw0d0d9xndb1rrxfnbk-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Scrumplex**
## Description of changes

Fixes #323262

### Notes for reviewers

The first commit does some general refactoring of the code. The `borgbackup` NixOS test derivation produced by this commit is identical to its parent commit.

The second commit contains the actual fixes.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Scrumplex** [comment]
Actually, I am just blind. There is a maintainer :D

File: nixos/modules/services/backup/borgbackup.nix [hunk 1]
```diff
@@ -20,8 +20,19 @@ let
     lib.concatStringsSep " "
       (lib.mapAttrsToList (x: y: "--keep-${x}=${toString y}") cfg.prune.keep);
 
+  mkExtraArgs = cfg:
+    # Create BASH arrays of extra args
+    lib.concatLines
+      (lib.mapAttrsToList (name: values: ''
+        ${name}=(${values})
```
Comment (Scrumplex) [inline nixos/modules/services/backup/borgbackup.nix:27]: After looking at this again, I think this should probably be to support spaces in arguments

```suggestion
        ${name}=(${lib.escapeShellArgs values})
```

Comment (Scrumplex) [reply nixos/modules/services/backup/borgbackup.nix:27]: actually, the options are already coerced with this function

File: nixos/modules/services/backup/borgbackup.nix [hunk 2]
```diff
@@ -20,8 +20,19 @@ let
...
+      '')
+  { inherit (cfg) extraArgs extraInitArgs extraCreateArgs extraPruneArgs extraCompactArgs; });
+
   mkBackupScript = name: cfg: pkgs.writeShellScript "${name}-script" (''
     set -e
+
+    ${mkExtraArgs cfg}
```
Comment (symphorien) [inline nixos/modules/services/backup/borgbackup.nix:34]: would it be possible to use structuredAttrs instead?

File: nixos/modules/services/backup/borgbackup.nix [hunk 3]
```diff
@@ -581,7 +591,7 @@ in {
             default = "";
             example = ''
               # To add excluded paths at runtime
-              extraCreateArgs="$extraCreateArgs --exclude /some/path"
```
Comment (symphorien) [inline nixos/modules/services/backup/borgbackup.nix:584]: as you had to change the example, it means that it's a breaking change worthy of a mentions in release notes, probably

File: nixos/tests/borgbackup.nix [hunk 4]
```diff
@@ -69,6 +71,7 @@ in {
             yearly = 5;
           };
           exclude = [ "*/${excludeFile}" ];
+          extraCreateArgs = [ "--exclude-caches" "--exclude-if-present" ".nobackup" ];
```
Comment (symphorien) [inline nixos/tests/borgbackup.nix:74]: I suppose to correctly test the feature the arguments should contain spaces or shell special characters

File: nixos/modules/services/backup/borgbackup.nix (hunk 2 above)
Comment (Scrumplex) [reply nixos/modules/services/backup/borgbackup.nix:34]: I think that would be great, but `writeShellScript` does not allow us to pass additional derivation args. We could use `writeTextFile` but then we would have to add a lot of boilerplate.

File: nixos/modules/services/backup/borgbackup.nix (hunk 3 above)
Comment (Scrumplex) [reply nixos/modules/services/backup/borgbackup.nix:584]: Agreed.

File: nixos/tests/borgbackup.nix (hunk 4 above)
Comment (Scrumplex) [reply nixos/tests/borgbackup.nix:74]: Agreed. I will change `.nobackup` to `.dont backup`

**Comment by necauqua** [comment]
Question relating to what I see in the release note:
Currently I have the workaround `extraCreateArgs=--exclude-caches` in the preHook

When I update to nixpkgs that has this merged, will it fail?
Requiring me to fix my config?

Or will it silently ignore the option, which in my case is not even as bad as if someone put their entire exclude list in the preHook - I ask just because I see this as something that can definitely happen with how ~~horrible~~ weird bash ~~is~~ can be about it and my lack of understanding of it.

**Comment by symphorien** [comment]
I think you have to replace that by
```
extraCreateArgs=(--exclude-caches)
```

**Comment by necauqua** [comment]
No I was not asking how to fix my setup (thanks though) - I was asking if I *dont* fix my setup will the thing just work, silently ignoring my incorrect config, or will it produce an explicit failure.

If someone had their exclude lists set with the `extraCreateArgs` and after this their now-incorrect args are ignored - this could be really bad, as their automated backup job will now pollute their repo with all sorts of garbage they wanted excluded, maybe not garbage but just a lot of gigabytes wasted, that was my concern.

Ideally for someone like me or that hypothetical person, after updating nixpkgs our config should fail to rebuild somehow, at least the service startup should fail (well it is part of the rebuild too).
If it just produces some error logs somewhere in systemd and continues *without* our args this could be bad, as I described above.

And my bash understanding is not enough to measure at a glance if my concern is valid or not :upside_down_face:

**Pull request merged**